In [8]:
!pip install lightgbm
import pandas as pd
import lightgbm as lgb


# 데이터 불러오기 (파일 이름 확인 필수)
train = pd.read_csv('/content/movies_train.csv')
test = pd.read_csv('/content/movies_test.csv')
submission = pd.read_csv('/content/submission.csv')


라이브러리 및 데이터 분석

In [9]:
df_train = pd.read_csv('movies_train.csv')
df_train.head()

,title,distributor,genre,release_time,time,screening_rat,director,dir_prev_bfnum,dir_prev_num,num_staff,num_actor,box_off_num
0,개들의 전쟁,롯데엔터테인먼트,액션,2012-11-22,96,청소년 관람불가,조병옥,NaN,0,91,2,23398
1,내부자들,(주)쇼박스,느와르,2015-11-19,130,청소년 관람불가,우민호,1161602.50,2,387,3,7072501
2,은밀하게 위대하게,(주)쇼박스,액션,2013-06-05,123,15세 관람가,장철수,220775.25,4,343,4,6959083
3,나는 공무원이다,(주)NEW,코미디,2012-07-12,101,전체 관람가,구자홍,23894.00,2,20,6,217866
4,불량남녀,쇼박스(주)미디어플렉스,코미디,2010-11-04,108,15세 관람가,신근호,1.00,1,251,2,483387


In [12]:
# df_train 데이터프레임의 상위 5개 행을 보여줍니다 (n의 기본값)
display(df_train.head())

,title,distributor,genre,release_time,time,screening_rat,director,dir_prev_bfnum,dir_prev_num,num_staff,num_actor,box_off_num
0,개들의 전쟁,롯데엔터테인먼트,액션,2012-11-22,96,청소년 관람불가,조병옥,NaN,0,91,2,23398
1,내부자들,(주)쇼박스,느와르,2015-11-19,130,청소년 관람불가,우민호,1161602.50,2,387,3,7072501
2,은밀하게 위대하게,(주)쇼박스,액션,2013-06-05,123,15세 관람가,장철수,220775.25,4,343,4,6959083
3,나는 공무원이다,(주)NEW,코미디,2012-07-12,101,전체 관람가,구자홍,23894.00,2,20,6,217866
4,불량남녀,쇼박스(주)미디어플렉스,코미디,2010-11-04,108,15세 관람가,신근호,1.00,1,251,2,483387


In [ ]:
# df_train 데이터프레임의 컬럼별 정보 (데이터 타입, Non-null 개수)를 출력합니다.
df_train.info()

In [13]:
# df_train 데이터프레임의 상위 10개 행을 보여줍니다 (n=10으로 지정)
display(df_train.head(10))

,title,distributor,genre,release_time,time,screening_rat,director,dir_prev_bfnum,dir_prev_num,num_staff,num_actor,box_off_num
0,개들의 전쟁,롯데엔터테인먼트,액션,2012-11-22,96,청소년 관람불가,조병옥,NaN,0,91,2,23398
1,내부자들,(주)쇼박스,느와르,2015-11-19,130,청소년 관람불가,우민호,1161602.50,2,387,3,7072501
2,은밀하게 위대하게,(주)쇼박스,액션,2013-06-05,123,15세 관람가,장철수,220775.25,4,343,4,6959083
3,나는 공무원이다,(주)NEW,코미디,2012-07-12,101,전체 관람가,구자홍,23894.00,2,20,6,217866
4,불량남녀,쇼박스(주)미디어플렉스,코미디,2010-11-04,108,15세 관람가,신근호,1.00,1,251,2,483387
5,강철대오 : 구국의 철가방,롯데엔터테인먼트,코미디,2012-10-25,113,15세 관람가,육상효,837969.00,2,262,4,233211
6,길위에서,백두대간,다큐멘터리,2013-05-23,104,전체 관람가,이창재,NaN,0,32,5,53526
7,회사원,(주)쇼박스,액션,2012-10-11,96,청소년 관람불가,임상윤,739522.00,3,342,2,1110523
8,"1789, 바스티유의 연인들",유니버설픽쳐스인터내셔널코리아,뮤지컬,2014-09-18,129,전체 관람가,정성복,NaN,0,3,5,4778
9,청춘그루브,(주)두타연,드라마,2012-03-15,94,15세 관람가,변성현,NaN,0,138,3,868


In [18]:
# df_train 데이터프레임의 마지막 5개 행을 tail() 메서드로 보여줍니다.
display(df_train.tail())

,title,distributor,genre,release_time,time,screening_rat,director,dir_prev_bfnum,dir_prev_num,num_staff,num_actor,box_off_num
595,해무,(주)NEW,드라마,2014-08-13,111,청소년 관람불가,심성보,3833.0,1,510,7,1475091
596,파파로티,(주)쇼박스,드라마,2013-03-14,127,15세 관람가,윤종찬,496061.0,1,286,6,1716438
597,살인의 강,(주)마운틴픽쳐스,공포,2010-09-30,99,청소년 관람불가,김대현,NaN,0,123,4,2475
598,악의 연대기,CJ 엔터테인먼트,느와르,2015-05-14,102,15세 관람가,백운학,NaN,0,431,4,2192525
599,베를린,CJ 엔터테인먼트,액션,2013-01-30,120,15세 관람가,류승완,NaN,0,363,5,7166532


In [20]:
display(df_train.shape)

(600, 12)

In [22]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           600 non-null    object 
 1   distributor     600 non-null    object 
 2   genre           600 non-null    object 
 3   release_time    600 non-null    object 
 4   time            600 non-null    int64  
 5   screening_rat   600 non-null    object 
 6   director        600 non-null    object 
 7   dir_prev_bfnum  270 non-null    float64
 8   dir_prev_num    600 non-null    int64  
 9   num_staff       600 non-null    int64  
 10  num_actor       600 non-null    int64  
 11  box_off_num     600 non-null    int64  
dtypes: float64(1), int64(5), object(6)
memory usage: 56.4+ KB


`df_train.info()`의 출력은 다음과 같은 정보를 제공합니다:

1.  **컬럼별 정보:** 각 컬럼의 이름이 나열됩니다.
2.  **비어있지 않은 (Non-Null) 값 개수:** 각 컬럼에 비어있지 않은 값(즉, 결측값이 아닌 값)의  개수를 보여줍니다.
*   title: 600개
*   distributor: 600개
*   genre: 600개
*   release_time: 600개
*   time: 600개
*   screening_rat: 600개
*   director: 600개
*   dir_prev_bfnum: 270개 (이 컬럼에 결측값이 존재합니다)
*   dir_prev_num: 600개
*   num_staff: 600개
*   num_actor: 600개
*   box_off_num: 600개
각 컬럼의 타입은 다음과 같습니다:

title, distributor, genre, release_time, screening_rat, director: object (주로 문자열)
time, dir_prev_num, num_staff, num_actor, box_off_num: int64 (정수)
dir_prev_bfnum: float64 (실수)

3.  **컬럼의 타입:** 각 컬럼의 데이터 타입을 (예: `int64`, `float64`, `object` 등) 표시합니다.

In [25]:
df_train['dir_prev_bfnum'] = df_train['dir_prev_bfnum'].fillna(0)
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           600 non-null    object 
 1   distributor     600 non-null    object 
 2   genre           600 non-null    object 
 3   release_time    600 non-null    object 
 4   time            600 non-null    int64  
 5   screening_rat   600 non-null    object 
 6   director        600 non-null    object 
 7   dir_prev_bfnum  600 non-null    float64
 8   dir_prev_num    600 non-null    int64  
 9   num_staff       600 non-null    int64  
 10  num_actor       600 non-null    int64  
 11  box_off_num     600 non-null    int64  
dtypes: float64(1), int64(5), object(6)
memory usage: 56.4+ KB


In [27]:
from sklearn.model_selection import KFold
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV

위 코드에서 `fillna(0)`를 사용하여 'dir_prev_bfnum' 컬럼의 모든 결측값을 0으로 대체했습니다. 이제 `df_train.info()`를 다시 실행하여 'dir_prev_bfnum' 컬럼의 Non-Null Count가 600으로 변경되었는지 확인해주세요.

In [29]:
display(df_train.describe())

,time,dir_prev_bfnum,dir_prev_num,num_staff,num_actor,box_off_num
count,600.000000,6.000000e+02,600.000000,600.000000,600.000000,6.000000e+02
mean,100.863333,4.726993e+05,0.876667,151.118333,3.706667,7.081818e+05
std,18.097528,1.309474e+06,1.183409,165.654671,2.446889,1.828006e+06
min,45.000000,0.000000e+00,0.000000,0.000000,0.000000,1.000000e+00
25%,89.000000,0.000000e+00,0.000000,17.000000,2.000000,1.297250e+03
50%,100.000000,0.000000e+00,0.000000,82.500000,3.000000,1.259100e+04
75%,114.000000,3.761416e+05,2.000000,264.000000,4.000000,4.798868e+05
max,180.000000,1.761531e+07,5.000000,869.000000,25.000000,1.426277e+07


In [30]:
# train과 test 데이터프레임을 합쳐서 전처리 과정을 일관되게 적용합니다.
df = pd.concat([train.drop('box_off_num', axis=1), test], ignore_index=True)

# release_time 컬럼을 datetime 객체로 변환합니다.
df['release_time'] = pd.to_datetime(df['release_time'])

# 연, 월, 일, 요일, 시간 피처를 추출합니다.
df['release_year'] = df['release_time'].dt.year
df['release_month'] = df['release_time'].dt.month
df['release_day'] = df['release_time'].dt.day
df['release_dayofweek'] = df['release_time'].dt.dayofweek
df['release_hour'] = df['release_time'].dt.hour

# 원본 release_time 컬럼은 더 이상 필요 없으므로 제거합니다.
df = df.drop('release_time', axis=1)

# dir_prev_bfnum 컬럼의 결측값을 0으로 채웁니다.
df['dir_prev_bfnum'] = df['dir_prev_bfnum'].fillna(0)

display(df.head())

,title,distributor,genre,time,screening_rat,director,dir_prev_bfnum,dir_prev_num,num_staff,num_actor,release_year,release_month,release_day,release_dayofweek,release_hour
0,개들의 전쟁,롯데엔터테인먼트,액션,96,청소년 관람불가,조병옥,0.00,0,91,2,2012,11,22,3,0
1,내부자들,(주)쇼박스,느와르,130,청소년 관람불가,우민호,1161602.50,2,387,3,2015,11,19,3,0
2,은밀하게 위대하게,(주)쇼박스,액션,123,15세 관람가,장철수,220775.25,4,343,4,2013,6,5,2,0
3,나는 공무원이다,(주)NEW,코미디,101,전체 관람가,구자홍,23894.00,2,20,6,2012,7,12,3,0
4,불량남녀,쇼박스(주)미디어플렉스,코미디,108,15세 관람가,신근호,1.00,1,251,2,2010,11,4,3,0


위 코드에서는 `train`과 `test` 데이터프레임을 합쳐 `release_time` 컬럼에서 연, 월, 일, 요일, 시간 피처를 추출했습니다. 또한 `dir_prev_bfnum` 컬럼의 결측값은 0으로 채웠습니다. 이제 범주형 변수를 처리할 차례입니다.

In [31]:
from sklearn.preprocessing import LabelEncoder

# 범주형 컬럼을 식별합니다.
categorical_features = ['distributor', 'genre', 'screening_rat', 'director']

# 각 범주형 컬럼에 대해 Label Encoding을 수행합니다.
for col in categorical_features:
    le = LabelEncoder()
    df[col] = le.fit_transform(df[col])

display(df.head())

,title,distributor,genre,time,screening_rat,director,dir_prev_bfnum,dir_prev_num,num_staff,num_actor,release_year,release_month,release_day,release_dayofweek,release_hour
0,개들의 전쟁,97,10,96,3,517,0.00,0,91,2,2012,11,22,3,0
1,내부자들,27,2,130,3,327,1161602.50,2,387,3,2015,11,19,3,0
2,은밀하게 위대하게,27,10,123,1,467,220775.25,4,343,4,2013,6,5,2,0
3,나는 공무원이다,7,11,101,2,26,23894.00,2,20,6,2012,7,12,3,0
4,불량남녀,122,11,108,1,282,1.00,1,251,2,2010,11,4,3,0


위 코드에서는 `distributor`, `genre`, `screening_rat`, `director`와 같은 범주형 컬럼에 `LabelEncoder`를 적용하여 수치형 데이터로 변환했습니다. 이제 데이터를 다시 `train`과 `test` 세트로 분리하고 모델 학습을 위한 준비를 할 수 있습니다.

In [33]:
# 전처리된 데이터를 다시 train과 test로 분리합니다.
X_train = df.iloc[:len(train)]
X_test = df.iloc[len(train):]
y_train = train['box_off_num']

display(X_train.head())
display(X_test.head())

,title,distributor,genre,time,screening_rat,director,dir_prev_bfnum,dir_prev_num,num_staff,num_actor,release_year,release_month,release_day,release_dayofweek,release_hour
0,개들의 전쟁,97,10,96,3,517,0.00,0,91,2,2012,11,22,3,0
1,내부자들,27,2,130,3,327,1161602.50,2,387,3,2015,11,19,3,0
2,은밀하게 위대하게,27,10,123,1,467,220775.25,4,343,4,2013,6,5,2,0
3,나는 공무원이다,7,11,101,2,26,23894.00,2,20,6,2012,7,12,3,0
4,불량남녀,122,11,108,1,282,1.00,1,251,2,2010,11,4,3,0


,title,distributor,genre,time,screening_rat,director,dir_prev_bfnum,dir_prev_num,num_staff,num_actor,release_year,release_month,release_day,release_dayofweek,release_hour
600,용서는 없다,133,2,125,3,147,3.005290e+05,2,304,3,2010,1,7,3,0
601,아빠가 여자를 좋아해,27,5,113,0,353,3.427002e+05,4,275,3,2010,1,14,3,0
602,하모니,60,4,115,0,1,4.206611e+06,3,419,7,2010,1,28,3,0
603,의형제,27,10,116,1,470,6.913420e+05,2,408,2,2010,2,4,3,0
604,평행 이론,60,1,110,1,38,3.173800e+04,1,380,1,2010,2,18,3,0


In [ ]:
import lightgbm as lgb

# LightGBM 모델 정의
model = lgb.LGBMRegressor(random_state=42)

# 모델 학습
model.fit(X_train, y_train)

# 예측 수행
predictions = model.predict(X_test)

display(predictions[:5])

위 코드에서는 `LGBMRegressor` 모델을 초기화하고, `X_train`과 `y_train`을 사용하여 모델을 학습시켰습니다. 학습된 모델은 `X_test` 데이터에 대한 예측을 생성했으며, 상위 5개의 예측값을 출력했습니다. 이제 예측 결과를 submission 파일 형식에 맞게 저장하겠습니다.

In [ ]:
# submission 파일에 예측 결과 저장
submission['box_off_num'] = predictions

# 예측값이 음수가 되는 경우 0으로 처리 (관객수는 음수가 될 수 없으므로)
submission['box_off_num'] = submission['box_off_num'].apply(lambda x: 0 if x < 0 else x)

# 결과 파일 저장
submission.to_csv('submission_lgbm.csv', index=False)

display(submission.head())

LightGBM 모델을 사용하여 예측을 완료하고, 예측 결과를 `submission_lgbm.csv` 파일로 저장했습니다. 예측값이 음수인 경우 0으로 처리하는 로직도 추가했습니다. 다음 단계는 K-Fold 교차 검증을 적용하거나 하이퍼파라미터 튜닝을 진행하는 것입니다.

In [ ]:
from sklearn.model_selection import KFold
import numpy as np

# K-fold 설정
NFOLDS = 5
फोल्डर = KFold(n_splits=NFOLDS, shuffle=True, random_state=42)

# K-fold 예측을 저장할 배열 초기화
kfold_predictions = np.zeros(X_test.shape[0])

# K-fold 모델 학습 및 예측
for fold_, (trn_idx, val_idx) in enumerate(फोल्डर.split(X_train, y_train)):
    print(f"Current Fold: {fold_}")
    trn_data = X_train.iloc[trn_idx]
    trn_target = y_train.iloc[trn_idx]

    val_data = X_train.iloc[val_idx]
    val_target = y_train.iloc[val_idx]

    # LightGBM 모델 정의 (기본 파라미터 사용)
    model = lgb.LGBMRegressor(random_state=42)

    # 모델 학습
    model.fit(trn_data, trn_target)

    # 예측 수행
    kfold_predictions += model.predict(X_test) / NFOLDS

display(kfold_predictions[:5])

위 코드에서는 5-겹 K-Fold 교차 검증을 사용하여 LightGBM 모델을 학습하고 예측을 수행했습니다. 각 폴드마다 모델을 훈련하고 `X_test`에 대한 예측을 생성한 후, 이 예측값들을 평균하여 최종 K-Fold 예측값을 도출했습니다. 이는 모델의 안정성과 일반화 성능을 높이는 데 도움이 됩니다.

이제 이 K-Fold 예측 결과를 submission 파일에 저장하겠습니다.

In [ ]:
# submission 파일에 K-Fold 예측 결과 저장
submission_kfold = submission.copy() # 원본 submission 데이터프레임을 복사하여 사용
submission_kfold['box_off_num'] = kfold_predictions

# 예측값이 음수가 되는 경우 0으로 처리
submission_kfold['box_off_num'] = submission_kfold['box_off_num'].apply(lambda x: 0 if x < 0 else x)

# 결과 파일 저장
submission_kfold.to_csv('submission_lgbm_kfold.csv', index=False)

display(submission_kfold.head())

K-Fold 교차 검증을 통해 얻은 예측값을 `submission_lgbm_kfold.csv` 파일로 저장했습니다. 다음 단계는 Grid Search를 통한 하이퍼파라미터 튜닝을 진행하거나, 추가적인 Feature Engineering을 고려해 볼 수 있습니다.

In [ ]:
from sklearn.model_selection import GridSearchCV

# LightGBM 모델 정의
lgbm = lgb.LGBMRegressor(random_state=42)

# 튜닝할 하이퍼파라미터 범위 설정
param_grid = {
    'n_estimators': [100, 200, 300],
    'learning_rate': [0.01, 0.05, 0.1],
    'num_leaves': [20, 31, 40],
    'max_depth': [-1, 10, 20]
}

# GridSearchCV 객체 생성
grid_search = GridSearchCV(estimator=lgbm, param_grid=param_grid, cv=KFold(n_splits=5, shuffle=True, random_state=42), scoring='neg_mean_squared_error', n_jobs=-1, verbose=1)

# 그리드 서치 수행
grid_search.fit(X_train, y_train)

# 최적의 하이퍼파라미터 및 최고 점수 출력
print("Best parameters found: ", grid_search.best_params_)
print("Lowest RMSE found: ", np.sqrt(-grid_search.best_score_))

위 코드에서는 `GridSearchCV`를 사용하여 LightGBM 모델의 최적 하이퍼파라미터를 탐색했습니다. `n_estimators`, `learning_rate`, `num_leaves`, `max_depth`에 대한 다양한 조합을 5-Fold 교차 검증을 통해 평가하고, 가장 낮은 RMSE를 달성하는 파라미터 조합을 찾았습니다.

이제 이 최적의 파라미터로 모델을 재학습하고 예측을 수행하겠습니다.

In [ ]:
# 최적의 하이퍼파라미터로 모델 재학습
best_lgbm_model = lgb.LGBMRegressor(**grid_search.best_params_, random_state=42)
best_lgbm_model.fit(X_train, y_train)

# 최적화된 모델로 예측 수행
optimized_predictions = best_lgbm_model.predict(X_test)

# submission 파일에 최적화된 예측 결과 저장
submission_optimized = submission.copy()
submission_optimized['box_off_num'] = optimized_predictions

# 예측값이 음수가 되는 경우 0으로 처리
submission_optimized['box_off_num'] = submission_optimized['box_off_num'].apply(lambda x: 0 if x < 0 else x)

# 결과 파일 저장
submission_optimized.to_csv('submission_lgbm_optimized.csv', index=False)

display(submission_optimized.head())

Grid Search를 통해 최적화된 LightGBM 모델로 예측을 수행하고, 그 결과를 `submission_lgbm_optimized.csv` 파일로 저장했습니다. 이로써 기본 LightGBM, K-Fold LightGBM, 그리고 하이퍼파라미터 튜닝을 거친 LightGBM 모델 학습 및 예측 과정이 모두 완료되었습니다.

모델 구축 및 검증 단계가 모두 완료되었습니다. 추가적으로 요청하실 사항이 있으신가요?

이제 전처리된 `X_train`, `X_test`와 타겟 변수 `y_train`이 준비되었습니다. 다음 단계는 LightGBM 모델을 구축하고 학습하는 것입니다.

`describe()` 메서드는 숫자형 컬럼에 대한 기술 통계(개수, 평균, 표준편차, 최소/최대값, 사분위수)를 요약하여 보여줍니다.